Support Scheme manifest

dangthaison91 opened the pull request:
This will resolve #195 and #199 

### Solution 📦
This PR allow user:
- Custom their Scheme and Tuist can generate more than 1 scheme for each target
- If there is no Scheme manifest for a target, auto-generate scheme for each target (keep current behavior)
- Add pre/post action for Build/Test/Archive Action

Usage:
``` swift
let TuistTarget = "Tuist-Demo"
let TuistTest = "Tuist-DemoTest"

let AppTarget = Target(name: TuistTarget,
                    platform: .iOS,
                    product: .app,
                    bundleId: "io.tuist.Tuist-Demo")

let TestTarget = Target(name: TuistTest,
                  platform: .iOS,
                  product: .unitTests,
                  bundleId: "io.tuist.Tests",
                  dependencies: [.target(name: TuistTarget)],
                  settings: Settings(base: ["TEST_HOST": "$(BUNDLE_LOADER)"]))

let DevScheme = Scheme(name: TuistTarget,
                       shared: true,
                       buildAction: BuildAction(targets: [TuistTarget, TuistTest],
                                                preActions: [ExecutionAction(title: "Dev", scriptText: "$ Dev script", target: TuistTarget)]),
                       testAction: TestAction(targets: [TuistTest]),
                       runAction: RunAction(executable: TuistTarget))

let ProductionScheme = Scheme(name: TuistTarget,
                              shared: true,
                              buildAction: BuildAction(targets: [TuistTarget, TuistTest],
                                                       preActions: [ExecutionAction(title: "Production", scriptText: "$ Production script", target: TuistTarget)]),
                              runAction: RunAction(executable: TuistTarget))

let project = Project(name: "Tuist-Demo",
                      up: [],
                      targets: [AppTarget, TestTarget],
                      schemes: [DevScheme, ProductionTarget])
```

### Implementation 👩‍💻👨‍💻

- [x] Update SchemeGenerator to support Scheme manifest and custom action
- [x] Update Tests
- [x] Update Changelog

ollieatkinson commented on Sources/TuistGenerator/Generator/SchemesGenerator.swift:
```diff
@@ -109,103 +154,117 @@ final class SchemesGenerator: SchemesGenerating {
             testables.append(testable)
         }
+        
+        scheme.testAction.flatMap { testAction in
```
this might read nicer as an `if let`

dangthaison91 replied:
Done, please check.

ollieatkinson commented on Sources/TuistGenerator/Generator/SchemesGenerator.swift:
```diff
@@ -109,103 +154,117 @@ final class SchemesGenerator: SchemesGenerating {
+                                                          projectName: generatedProject.name)
+            
+            let testable = XCScheme.TestableReference(skipped: false, buildableReference: reference)
             testables.append(testable)
```
Generally if there is a loop and inside the closure for the loop contents are being appended to an array it might make a good use-case for the `map` function. Maybe worth considering here - it's a very minor point so feel free to ignore it.

dangthaison91 replied:
As you see, I'm using `guard let` above therefore using `map` will require return `nil` then filter nil in the end.

I also want to keep coding style as before.

ollieatkinson commented on Sources/TuistGenerator/Generator/SchemesGenerator.swift:
```diff
@@ -109,103 +154,117 @@ final class SchemesGenerator: SchemesGenerating {
+            entries.append(XCScheme.BuildAction.Entry(buildableReference: buildableReference, buildFor: buildFor))
+        }
 
+        scheme.buildAction.flatMap { buildAction in
```
Same points as above.

dangthaison91 commented on Sources/ProjectDescription/Scheme.swift:
```diff
@@ -30,6 +30,27 @@ public class Scheme: Codable {
+    
+    public enum CodingKeys: String, CodingKey {
+        case title
+        case scriptText = "script_text"
```
As I see, we manually specify coding keys for several class: Target, Scheme... Is it necessary to do it explicitly if we want to have snake_case keys? I don't see we define `keyEncodingStrategy = .convertToSnakeCase` anywhere?

dangthaison91 commented on Sources/TuistGenerator/Models/Scheme.swift:
```diff
@@ -57,25 +57,77 @@ public class Arguments: Equatable {
+    }
+}
+
+public class SerialAction: Equatable {
```
I think duplicate 2 properties is better than using inheritance here. `SerialAction` is removed.

dangthaison91 commented on Sources/TuistGenerator/Models/Scheme.swift:
```diff
@@ -57,25 +57,77 @@ public class Arguments: Equatable {
+    
+    // MARK: - Equatable
+    
+    public static func == (lhs: SerialAction, rhs: SerialAction) -> Bool {
```
Thanks @kwridan , `SerialAction` is removed.

dangthaison91 commented on Tests/TuistIntegrationTests/Generator/MultipleConfigurationsIntegrationTests.swift:
```diff
@@ -284,12 +284,13 @@ final class MultipleConfigurationsIntegrationTests: XCTestCase {
     }
 
-    private func createProject(path: AbsolutePath, settings: Settings, targets: [Target]) -> Project {
+    private func createProject(path: AbsolutePath, settings: Settings, targets: [Target], schemes: [Scheme]) -> Project {
```
If scheme is not shared, it will be generated at path : `username.xcuserdatad/schemes`. Do you have a any ideas how we can get Mac OS username?
> **shared** flag within Scheme doesn't seem to toggle the setting in the generated scheme

Updated test coverage setting for TestAction
> **coverage** flag within TestAction doesn't seem to enable test coverage in the generated project

dangthaison91 commented:
I just update PR to support **Launch arguments** and **Environment Variables** @ollieatkinson

dangthaison91 commented:
@kwridan 
If scheme is not shared, it will be generated at path : `username.xcuserdatad/schemes`. Do you have a any ideas how we can get Mac OS username?
> **shared** flag within Scheme doesn't seem to toggle the setting in the generated scheme

Updated test coverage setting for TestAction
> **coverage** flag within TestAction doesn't seem to enable test coverage in the generated project

I also added fixture `ios_app_with_custom_scheme`
> Can we please add a fixture 🙂 - ios_app_with_custom_schemes this will aid testing out this feature.

dangthaison91 commented:
Should I update CHANGELOG only and update document in another PR @pepibumur @kwridan ?

dangthaison91 commented:
I updated PR to follow your suggestions @kwridan . Ready to merge.

kwridan commented:
👍

pepibumur commented:
Great contribution @dangthaison91 !
![giphy](https://user-images.githubusercontent.com/663605/57512876-2d1aa200-730d-11e9-935e-a65fbbdc42b6.gif)
